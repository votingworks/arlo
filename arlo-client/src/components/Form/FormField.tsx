import React from 'react'
import styled from 'styled-components'
import { InputGroup, NumericInput } from '@blueprintjs/core'
import { getIn, FieldProps } from 'formik'

const Wrapper = styled.div`
  width: 45%;
`

const Field = styled(InputGroup)`
  width: 100%;
`

const NumberField = styled(NumericInput)`
  width: 100%;
`

const ErrorLabel = styled.p`
  width: 100%;
  color: #ff0000;
`

interface Props {
  field: FieldProps['field']
  form: FieldProps['form']
  disabled?: boolean
  value?: string | number
  onChange?: (e: React.ChangeEvent) => void
  onBlur?: (e: React.FocusEvent) => void
  name?: string
  type?: string
  error?: string
  touched?: boolean
  className?: string
}

const FormField: React.FC<Props> = ({
  field,
  form: { touched, errors, setFieldTouched, setFieldValue },
  disabled,
  className,
  ...rest
}: Props) => (
  <Wrapper className={className}>
    {rest.type === 'number' ? (
      <NumberField
        disabled={disabled}
        onValueChange={(n, s) => {
          setFieldValue(field.name, n)
          setFieldTouched(field.name)
        }}
        {...field}
        {...rest}
      />
    ) : (
      <Field
        disabled={disabled}
        {...field}
        {...rest}
        onChange={field.onChange}
      />
    )}
    {getIn(errors, field.name) && getIn(touched, field.name) && (
      <ErrorLabel data-testid={`${field.name}-error`}>
        {getIn(errors, field.name)}
      </ErrorLabel>
    )}
  </Wrapper>
)

export default FormField
