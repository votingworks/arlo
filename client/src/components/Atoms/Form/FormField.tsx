import React from 'react'
import styled from 'styled-components'
import { InputGroup, NumericInput, TextArea } from '@blueprintjs/core'
import { getIn, FieldProps } from 'formik'

interface IWrapperProps {
  wide: boolean
}

const Wrapper = styled.div<IWrapperProps>`
  width: ${p => (p.wide ? '100%' : '45%')};
`

const Field = styled(InputGroup)`
  margin-top: 5px;
  width: 100%;
`

const Area = styled(TextArea)`
  margin-top: 5px;
  width: 100%;
`

const NumberField = styled(NumericInput)`
  margin-top: 5px;
  width: 100%;

  .bp3-input-group {
    width: 100%;
  }

  input.bp3-input {
    margin-top: 0;
  }

  .bp3-button-group.bp3-vertical.bp3-fixed {
    transform: translateX(-100%);
    z-index: 15;
  }
`

const ErrorLabel = styled.p`
  width: 100%;
  color: #ff0000;
`

export interface IProps {
  field: FieldProps['field']
  form: Pick<
    FieldProps['form'],
    'touched' | 'errors' | 'setFieldTouched' | 'setFieldValue'
  >
  disabled?: boolean
  value?: string | number
  onChange?: (e: React.FormEvent) => void
  onBlur?: (e: React.FocusEvent) => void
  name?: string
  type?: string
  error?: string
  touched?: boolean
  className?: string
}

const FormField: React.FC<IProps> = ({
  field,
  form: { touched, errors, setFieldTouched, setFieldValue },
  disabled,
  className,
  type,
  ...rest
}: IProps) => (
  <Wrapper className={className} wide={type === 'textarea'}>
    {type === 'number' ? (
      <NumberField
        disabled={disabled}
        onValueChange={n => setFieldValue(field.name, n)}
        type={type}
        {...field}
        {...rest}
        onBlur={() => setFieldTouched(field.name)}
      />
    ) : type === 'textarea' ? (
      <Area disabled={disabled} onChange={rest.onChange} {...field} {...rest} />
    ) : (
      <Field disabled={disabled} type={type} {...field} {...rest} />
    )}
    {getIn(errors, field.name) && getIn(touched, field.name) && (
      <ErrorLabel data-testid={`${field.name}-error`}>
        {getIn(errors, field.name)}
      </ErrorLabel>
    )}
  </Wrapper>
)

export default FormField
