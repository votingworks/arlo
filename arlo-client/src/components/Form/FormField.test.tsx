import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { FormikProps } from 'formik'
import FormField from './FormField'

describe('FormField', () => {
  it('renders a Field', () => {
    const field = {
      name: 'field-name',
      value: 'text value',
      onChange: jest.fn(),
      onBlur: jest.fn(),
    }
    const form = ({
      errors: {},
      touched: {},
    } as any) as FormikProps<any> // eslint-disable-line @typescript-eslint/no-object-literal-type-assertion
    const { container } = render(
      <FormField field={field} form={form} type="text" />
    )

    expect(container).toMatchSnapshot()
  })

  it('renders a Field with errors', () => {
    const field = {
      name: 'field-name',
      value: '',
      onChange: jest.fn(),
      onBlur: jest.fn(),
    }
    const form = ({
      errors: {
        'field-name': 'Required',
      },
      touched: {
        'field-name': true,
      },
    } as any) as FormikProps<any> // eslint-disable-line @typescript-eslint/no-object-literal-type-assertion
    const { container } = render(
      <FormField field={field} form={form} type="text" />
    )

    expect(container).toMatchSnapshot()
  })

  it('calls onChange', () => {
    const field = {
      name: 'field-name',
      value: '',
      onChange: jest.fn(),
      onBlur: jest.fn(),
    }
    const form = ({
      errors: {},
      touched: {},
    } as any) as FormikProps<any>
    const { getByTestId } = render(
      <FormField field={field} form={form} type="text" data-testid="testid" />
    )

    fireEvent.change(getByTestId('testid'), { target: { value: 'new value' } })

    expect(field.onChange).toHaveBeenCalledTimes(1)
  })
})
