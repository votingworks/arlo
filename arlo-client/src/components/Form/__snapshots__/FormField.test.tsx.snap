// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField renders a Field 1`] = `
<div>
  <div
    class="sc-bdVaJa NQGah"
  >
    <div
      class="bp3-input-group sc-bwzfXH jwfMIB"
    >
      <input
        class="bp3-input"
        name="field-name"
        style="padding-right: 10px;"
        type="text"
        value="text value"
      />
    </div>
  </div>
</div>
`;

exports[`FormField renders a Field with errors 1`] = `
<div>
  <div
    class="sc-bdVaJa NQGah"
  >
    <div
      class="bp3-input-group sc-bwzfXH jwfMIB"
    >
      <input
        class="bp3-input"
        name="field-name"
        style="padding-right: 10px;"
        type="text"
        value=""
      />
    </div>
    <p
      class="sc-bxivhb hSbDWx"
      data-testid="field-name-error"
    >
      Required
    </p>
  </div>
</div>
`;
